lib/ClangImporter/ClangImporter.cpp

```diff
@@ -4005,6 +4006,578 @@ void ClangImporter::Implementation::lookupAllObjCMembers(
+
+  return pointeePropertyRefExpr;
+}
+
+// For const methods generate the following:
+//   %0 = __swift_interopStaticCast<Base>(self)
+//   return %0.fn(args...)
+// For mutating methods we have to pass self as a pointer:
+//   %0 = Builtin.addressof(&self)
+//   %1 = Builtin.reinterpretCast<UnsafeMutablePointer<Derived>>(%0)
+//   %2 = __swift_interopStaticCast<UnsafeMutablePointer<Base>?>(%1)
+//   %3 = %2!
+//   %4 = %3.pointee
+//   return %4.fn(args...)
+static std::pair<BraceStmt *, bool>
+synthesizeBaseClassMethodBody(AbstractFunctionDecl *afd, void *context) {
+  ASTContext &ctx = afd->getASTContext();
+
+  auto funcDecl = cast<FuncDecl>(afd);
+  auto derivedStruct =
+      cast<StructDecl>(funcDecl->getDeclContext()->getAsDecl());
+  auto baseMember = static_cast<FuncDecl *>(context);
+  auto baseStruct = cast<StructDecl>(baseMember->getDeclContext()->getAsDecl());
+  auto baseType = baseStruct->getDeclaredType();
+
+  SmallVector<Expr *, 8> forwardingParams;
+  for (auto param : *funcDecl->getParameters()) {
+    auto paramRefExpr = new (ctx) DeclRefExpr(param, DeclNameLoc(),
+                                              /*Implicit=*/true);
+    paramRefExpr->setType(param->getType());
+    forwardingParams.push_back(paramRefExpr);
+  }
+
+  Expr *casted = nullptr;
+  if (funcDecl->isMutating()) {
+    auto pointeeMemberRefExpr =
+        getInOutSelfInteropStaticCast(funcDecl, baseStruct, derivedStruct);
+    casted = new (ctx) InOutExpr(SourceLoc(), pointeeMemberRefExpr, baseType,
+                                 /*implicit*/ true);
+    casted->setType(InOutType::get(baseType));
+  } else {
+    auto *selfDecl = funcDecl->getImplicitSelfDecl();
+    auto selfExpr = new (ctx) DeclRefExpr(selfDecl, DeclNameLoc(),
+                                          /*implicit*/ true);
+    selfExpr->setType(selfDecl->getType());
+
+    auto staticCastRefExpr = getInteropStaticCastDeclRefExpr(
+        ctx, baseStruct->getClangDecl()->getOwningModule(), baseType,
+        derivedStruct->getDeclaredType());
+
+    auto castedCall = CallExpr::createImplicit(ctx, staticCastRefExpr,
+                                               {selfExpr}, {Identifier()});
+    castedCall->setType(baseType);
+    castedCall->setThrows(false);
+    casted = castedCall;
+  }
+
+  auto *baseMemberExpr =
+      new (ctx) DeclRefExpr(ConcreteDeclRef(baseMember), DeclNameLoc(),
+                            /*Implicit=*/true);
+  baseMemberExpr->setType(baseMember->getInterfaceType());
+
+  auto baseMemberDotCallExpr =
+      DotSyntaxCallExpr::create(ctx, baseMemberExpr, SourceLoc(), casted);
+  baseMemberDotCallExpr->setType(baseMember->getMethodInterfaceType());
+  baseMemberDotCallExpr->setThrows(false);
+
+  auto *baseMemberCallExpr = CallExpr::createImplicit(
+      ctx, baseMemberDotCallExpr, forwardingParams, {});
+  baseMemberCallExpr->setType(baseMember->getResultInterfaceType());
+  baseMemberCallExpr->setThrows(false);
```

Comment:
Could this be turned into one call?

Reply:
After some investigation, no. 

Essentially member functions are imported with the following signature `(Self) -> () -> Void`. So the first expression gets us our actual function by passing self in. Then we need to call the actual function, which is the result of the first function.